sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -965,12 +968,33 @@ class RocksDB(
 
   /**
    * Function to release RocksDB instance lock that allows for synchronized access to the state
-   * store instance
+   * store instance. Optionally provide a thread to check against, and release only if provided
+   * thread is the one that acquired the lock.
    *
    * @param opType - operation type releasing the lock
+   * @param releaseForThreadOpt - optional thread to check against acquired thread
    */
-  private def release(opType: RocksDBOpType): Unit = acquireLock.synchronized {
+  private def release(
+      opType: RocksDBOpType,
+      releaseForThreadOpt: Option[AcquiredThreadInfo] = None): Unit = acquireLock.synchronized {
     if (acquiredThreadInfo != null) {
+      if (releaseForThreadOpt.nonEmpty) {
+        if (releaseForThreadOpt.get.threadRef.get.isEmpty) {
+          logInfo(log"Thread reference is empty when attempting to release for" +
+            log" opType=${MDC(LogKeys.OP_TYPE, opType.toString)}, ignoring release." +
+            log" Lock is held by ${MDC(LogKeys.THREAD, acquiredThreadInfo)}")
+          return
+        }
+
+        if (acquiredThreadInfo.threadRef.get.isDefined
+          && acquiredThreadInfo != releaseForThreadOpt.get) {
```

Comment:
Question: in `acquire()` method, we compare by comparing the threadId, i.e.
`newAcquiredThreadInfo.threadRef.get.get.getId != acquiredThreadInfo.threadRef.get.get.getId`

But here we are comparing using the entire `ThreadInfo`. Trying to understand why? Why are we not doing this in `acquire`? Is there a reason? Ideally, we should use a consistent approach. cc @anishshri-db

Reply:
I think we should move away from this model anyway for the long term. but sure, for now - we can keep the comparisons consistent across acquire and release

Reply:
@liviazhu-db I agree that comparing the entire object is a better pattern, my question here was more about why they are different, just to make sure we know why and not missing something. Don't want us to be consistent for just consistency sake. If we believe comparing the full object is the safer approach then lets do that. We can just do that for only this release func for now.